pkg/evaluator/evaluator.go

```diff
@@ -213,6 +215,94 @@ func (e *Evaluator) evalWhile(scope *scope, w *parser.While) Value {
 	return val
 }
 
+func (e *Evaluator) evalFor(scope *scope, f *parser.For) Value {
+	scope = newInnerScope(scope)
+	r, err := e.newRange(scope, f)
+	if err != nil {
+		return err
+	}
+	for r.next() {
+		val := e.Eval(scope, f.Block)
+		if isError(val) || isBreak(val) || isReturn(val) {
+			return val
+		}
+	}
+	return nil
+}
+
+func (e *Evaluator) newRange(scope *scope, f *parser.For) (ranger, Value) {
+	if r, ok := f.Range.(*parser.StepRange); ok {
+		return e.newStepRange(scope, r, f.LoopVar)
+	}
+	rangeVal := e.Eval(scope, f.Range)
+	if isError(rangeVal) {
+		return nil, rangeVal
+	}
+
+	switch v := rangeVal.(type) {
+	case *Array:
+		loopVar := zero(f.LoopVar.Type())
+		scope.set(f.LoopVar.Name, loopVar)
+		return &arrayRange{loopVar: loopVar, array: v, cur: 0}, nil
+	case *String:
+		loopVar := &String{}
+		scope.set(f.LoopVar.Name, loopVar)
+		return &stringRange{loopVar: loopVar, str: v, cur: 0}, nil
+	case *Map:
+		loopVar := &String{}
+		scope.set(f.LoopVar.Name, loopVar)
+		order := make([]string, len(*v.Order))
+		copy(order, *v.Order)
```

Comment:
Don't forget to document this - that iteration over a map only iterates over the entries present at the start, and not new ones added. And that it is safe to delete entries.

Reply:
yeah... there were some docs updates, but there is already a doc update in flight. should get that merged...

Reply:
yeah... there were some docs updates, but there is already a doc update in flight. should get that merged maybe and rework it again ;)

Reply:
I've added a note to my longer term todos here:
https://gist.github.com/juliaogris/05d459cbdb475c53cced8c06dde71072